### ISSUE TITLE
Internal error when LET variable conflicts with collection name

### ISSUE THREAD

**Issue by DeShadow** [arangodb/arangodb#2168]
ArangoDB 3.1

Simple query:
```
let users = [{ a: '123' }]

let u = document(users, '1')
return u
```

throws internal error
```
Query: internal error (exception location: /var/lib/jenkins/workspace/ArangoDB_Release/arangod/Aql/Functions.cpp:2702). Please report this error to arangodb.com (while executing) (exception location: /var/lib/jenkins/workspace/ArangoDB_Release/arangod/RestHandler/RestCursorHandler.cpp:131). Please report this error to arangodb.com
```

I have collection users, but I can declare variable with the same name. In this case query ends with internal error.

**Comment by jsteemann** [comment]
Also affects 3.0.

**Comment by jsteemann** [comment]
Potential fix:
```diff
diff --git a/arangod/Aql/Functions.cpp b/arangod/Aql/Functions.cpp
index de4f8349..cda9587 100644
--- a/arangod/Aql/Functions.cpp
+++ b/arangod/Aql/Functions.cpp
@@ -2502,7 +2502,9 @@ AqlValue Functions::Document(arangodb::aql::Query* query,
 
   AqlValue collectionValue = ExtractFunctionParameterValue(trx, parameters, 0);
   if (!collectionValue.isString()) {
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
+    RegisterWarning(query, "DOCUMENT",
+                    TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH);
+      return AqlValue(arangodb::basics::VelocyPackHelper::NullValue());
   }
   std::string collectionName(collectionValue.slice().copyString());
 
```

**Comment by Simran-B** [comment]
```js
LET Users = {foo: "bar"}
RETURN DOCUMENT(Users, "A")
```

is turned into

```js
RETURN DOCUMENT({ "foo" : "bar" }, "A")
```

by the optimizer. It does not matter if a collection `Users` is present, an internal error raised either way.

Without the `, "A"` part, `[ null ]` is returned.

```js
RETURN DOCUMENT({ _id: "Users/A" })
```

does not find the document either and returns `[ null ]`. I would say that this is expected, as no objects should be passed to the DOCUMENT() function, but rather strings (e.g. `DOCUMENT("Users", "A")` or `DOCUMENT("Users/A")`).

An internal error should not occur however, I would instead expect an early query error because of bad argument types in the function call, as @jsteemann suggests above.

**Comment by Simran-B** [comment]
```js
LET Users = {foo: "bar"}
RETURN V8(DOCUMENT(Users, "A"))
```

returns `[ null ]` BTW, no error and also no warning. I guess the JavaScript implementation should behave the same as the C++ version, so both should issue an argument type mismatch warning.

**Comment by DeShadow** [comment]
@Simran-B Yeah, you are right! :) I haven't tested this issue much time, that's why I posted it 'as is'.

**Comment by jsteemann** [comment]
Fixed in 3.0.12 and 3.1.1.

**Comment by dothebart** [comment]
3.0.12 and 3.1.1 are available for download via the regular upgrade channels - closing.
